File: frontend/src/types/todo.ts
import { z } from "zod";

export type Todo = {
  id: number;
  title: string;
  description: string;
  completed: Date;
  userId: number;
};
export const todoCreateValidator = z.object({
  title: z.string().min(3).max(32),
  description: z.string().max(255).optional(),
});
export type TodoCreate = z.infer<typeof todoCreateValidator>;

export type TodoUpdate = Partial<TodoCreate> & { id: number };

export type TodoSearch = {
  showComplete: boolean;
};

File: backend/models/todo.py
from datetime import datetime
from typing import Optional

from pydantic import Field, BaseModel
from generated.prisma.bases import BaseTodo


class TodoCreateInput(BaseModel):
    title: str = Field(..., description="Title of the todo")
    description: str | None = Field(
        None, nullable=True, description="Description of the todo"
    )


class TodoUpdateInput(BaseModel):
    title: str | None = Field(None, nullable=True, description="Title of the todo")
    description: str | None = Field(
        None, nullable=True, description="Description of the todo"
    )


class Todo(BaseModel):
    id: int = Field(..., description="Id of the todo")
    title: str = Field(..., description="Title of the todo")
    description: str | None = Field(None, nullable=True)
    userId: int = Field(..., description="Id of the user who created the todo")
    createdAt: datetime = Field(..., description="Date the todo was created")
    updatedAt: datetime = Field(..., description="Date the todo was last updated")
    completed: datetime | None = Field(
        None, nullable=True, description="Date the todo was completed"
    )


class TodoSearch(BaseModel):
    showIncomplete: bool | None = Field(
        False, nullable=True, description="Whether to show incomplete todos"
    )
